src/hpa.c

```diff
@@ -840,21 +857,73 @@ hpa_dalloc_batch(tsdn_t *tsdn, pai_t *self, edata_list_active_t *list) {
 		hpa_dalloc_locked(tsdn, shard, edata);
 	}
 	hpa_shard_maybe_do_deferred_work(tsdn, shard, /* forced */ false);
+	*deferred_work_generated =
+	    hpa_shard_has_deferred_work(tsdn, shard);
+
 	malloc_mutex_unlock(tsdn, &shard->mtx);
 }
 
 static void
-hpa_dalloc(tsdn_t *tsdn, pai_t *self, edata_t *edata) {
+hpa_dalloc(tsdn_t *tsdn, pai_t *self, edata_t *edata,
+    bool *deferred_work_generated) {
 	/* Just a dalloc_batch of size 1; this lets us share logic. */
 	edata_list_active_t dalloc_list;
 	edata_list_active_init(&dalloc_list);
 	edata_list_active_append(&dalloc_list, edata);
-	hpa_dalloc_batch(tsdn, self, &dalloc_list);
+	hpa_dalloc_batch(tsdn, self, &dalloc_list, deferred_work_generated);
 }
 
+/*
+ * Calculate time until either purging or hugification ought to happen.
+ * Called by background threads.
+ */
 static uint64_t
 hpa_time_until_deferred_work(tsdn_t *tsdn, pai_t *self) {
```

Comment:
Hmmm, this gets called after most operations, right? I would really rather avoid acquiring the lock a second time again if so.

Instead, we can update the "time until next deferred operation is ready" on every function call, and have a lock-free way of reading it in this function. (Probably using a seqlock, as in seq.h, since we don't have a guarantee that 64-bit atomics are available on every platform).

Reply:
It's called only by background thread after it's done with some background work or woke up from indefinite sleep. Do you think it will significantly contribute to the contention?

Reply:
OK, I think that makes sense; that case should be rare.